File: Data-Structures-Algorithms/Math/Digit_Conversions/roman-to-integer.cpp
/*
https://leetcode.com/problems/roman-to-integer/
*/

# include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int romanToInt(string s) {
        
        unordered_map<char, int> roman_to_int = {{'I', 1}, {'V', 5}, {'X', 10}, {'L', 50}, {'C', 100}, {'D', 500}, {'M', 1000}};
        
        int s_len = s.length();
        int i, int_total = 0;
        char curr, prev;
        
        prev = s[0];
        i = 1;
        
        while(i < s_len) {
            
            curr = s[i];
            
            if ((prev == 'I' and (curr == 'V' or curr == 'X')) or (prev == 'X' and (curr == 'L' or curr == 'C')) or (prev == 'C' and (curr == 'D' or curr == 'M')))
                int_total -= roman_to_int[prev];
            else
                int_total += roman_to_int[prev];
            
            prev = curr;
            
            i++;
            
        }
                                                                   
        int_total += roman_to_int[s[s_len - 1]];
                                                                   
        return int_total;                
    }

    void solve() {
        
        string roman_numeral;
        cin >> roman_numeral;
        cout << romanToInt(roman_numeral) << endl;

    }
};

int main() {

    /*
    MCMXCIV
    */

    /*
    1994
    */

    Solution *soln = new Solution;
    string roman_numeral;
    cin >> roman_numeral;
    cout << soln -> romanToInt(roman_numeral) << endl;
    
    return 0;
}

File: Data-Structures-Algorithms/Math/Digit-Conversions/roman-to-integer.java

    import java.util.HashMap;
import java.util.Scanner;

public class Solution {
    public int romanToInt(String s) {
        HashMap<Character, Integer> romanToInt = new HashMap<>();
        romanToInt.put('I', 1);
        romanToInt.put('V', 5);
        romanToInt.put('X', 10);
        romanToInt.put('L', 50);
        romanToInt.put('C', 100);
        romanToInt.put('D', 500);
        romanToInt.put('M', 1000);

        int sLen = s.length();
        int intTotal = 0;

        char prev = s.charAt(0);
        int i = 1;

        while (i < sLen) {
            char curr = s.charAt(i);

            if ((prev == 'I' && (curr == 'V' || curr == 'X')) || (prev == 'X' && (curr == 'L' || curr == 'C')) ||
                    (prev == 'C' && (curr == 'D' || curr == 'M'))) {
                intTotal -= romanToInt.get(prev);
            } else {
                intTotal += romanToInt.get(prev);
            }

            prev = curr;
            i++;
        }

        intTotal += romanToInt.get(s.charAt(sLen - 1));

        return intTotal;
    }

    public void solve() {
        Scanner scanner = new Scanner(System.in);
        String romanNumeral = scanner.next();
        System.out.println(romanToInt(romanNumeral));
    }

    public static void main(String[] args) {
        Solution soln = new Solution();
        Scanner scanner = new Scanner(System.in);
        String romanNumeral = scanner.next();
        System.out.println(soln.romanToInt(romanNumeral));
    }
}

